### PR TITLE
Log internal errors encountered in typeinfer

### PR DESCRIPTION
**Pull request by sklam**
instead of reporting it since the report is not readable anyway

TODOS:
- [x] trying it more to see if this is really useful
- [x] add options to make runtests enable logging

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Sorry @sklam, accidentally hit enter and closed the PR!

**Comment by sklam** [comment]
I forgot that the cuda driver is already using logging.

**Comment by seibert** [comment]
Do we want to also turn on logging in our Azure build scripts with this PR?

**Comment by stuartarchibald** [comment]
Also, should the e.g. developer docs have an entry on enabling logging?

File: numba/typeinfer.py [hunk 1]
```diff
@@ -140,17 +143,20 @@ def propagate(self, typeinfer):
...
                     errors.append(e)
-                except Exception:
-                    msg = "Internal error at {con}:\n{sep}\n{err}{sep}\n"
-                    e = TypingError(msg.format(con=constraint,
-                                               err=traceback.format_exc(),
-                                               sep='--%<' + '-' * 76),
-                                    loc=constraint.loc,
-                                    highlighting=False)
+                except Exception as e:
+                    _logger.exception("captured error")
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/typeinfer.py]: I find it useful for debugging.  Note the `.exception` method will log the active exception+traceback.

[the remark this answers is not in the document]
Comment (sklam) [reply numba/typeinfer.py]: I just realized that the `.exception` will log at `ERROR` level, which is above the default threshold; thus, it is always showing the traceback.  It is too noisey.  I am switching it to a `.debug`

File: numba/typeinfer.py [hunk 2]
```diff
@@ -140,17 +143,20 @@ def propagate(self, typeinfer):
                 try:
                     constraint(typeinfer)
                 except TypingError as e:
+                    _logger.exception("captured error")
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/typeinfer.py]: same as above

**Comment by sklam** [comment]
I added docs in the "Running tests" session about the `--log` flag.

**Comment by sklam** [comment]
(Addressing https://github.com/numba/numba/pull/3646#pullrequestreview-205258679) 
By switching from `logger.exception` to `logger.debug`, the traceback is back to normal now.

File: numba/typeinfer.py [hunk 3]
```diff
@@ -143,17 +147,20 @@ def propagate(self, typeinfer):
...
-                    msg = "Internal error at {con}:\n{sep}\n{err}{sep}\n"
-                    e = TypingError(msg.format(con=constraint,
-                                               err=traceback.format_exc(),
-                                               sep='--%<' + '-' * 76),
-                                    loc=constraint.loc,
-                                    highlighting=False)
+                except Exception as e:
+                    _logger.debug("captured error", exc_info=e)
+                    msg = ("Internal error at {con}.\n"
+                           "{err}\nEnable logging for details.")
```
Comment (stuartarchibald) [inline numba/typeinfer.py]:
```suggestion
                           "{err}\nEnable logging at debug level for details.")
```

**Comment by stuartarchibald** [comment]
Thanks for the fixes, one minor suggested change to the wording and then I think this is good to mark for merge.

**Pull request merged**